File: src/api/src/weckpi/api/config/alarm_config.py
"""All alarm related configuration models."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import time
from typing import Literal


@dataclass
class AlarmConfig:
    """
    The root model for the alarm system config.

    :var monday: Config for the alarm on monday.
    :var tuesday: Config for the alarm on tuesday.
    :var wednesday: Config for the alarm on wednesday.
    :var thursday: Config for the alarm on thursday.
    :var friday: Config for the alarm on friday.
    :var saturday: Config for the alarm on saturday.
    :var sunday: Config for the alarm on sunday.
    """
    monday: AlarmDay
    tuesday: AlarmDay
    wednesday: AlarmDay
    thursday: AlarmDay
    friday: AlarmDay
    saturday: AlarmDay
    sunday: AlarmDay

    def get_alarm_config(self, iso_weekday: Literal[1, 2, 3, 4, 5, 6, 7]) -> AlarmDay:
        """Get the alarm config for the number of the weekday (monday = 1, sunday = 7)."""
        day_mapping = {
            1: 'monday',
            2: 'tuesday',
            3: 'wednesday',
            4: 'thursday',
            5: 'friday',
            6: 'saturday',
            7: 'sunday'
        }
        return getattr(self, day_mapping[iso_weekday])

    @classmethod
    def from_json(cls, config_json: dict) -> AlarmConfig:
        """Create a AlarmConfig object from a dict."""
        return cls(
            AlarmDay.from_json(config_json['monday']),
            AlarmDay.from_json(config_json['tuesday']),
            AlarmDay.from_json(config_json['wednesday']),
            AlarmDay.from_json(config_json['thursday']),
            AlarmDay.from_json(config_json['friday']),
            AlarmDay.from_json(config_json['saturday']),
            AlarmDay.from_json(config_json['sunday'])
        )

    def to_json(self) -> dict:
        """Dump this AlarmConfig object into a dict."""
        return {
            'monday': self.monday.to_json(),
            'tuesday': self.tuesday.to_json(),
            'wednesday': self.wednesday.to_json(),
            'thursday': self.thursday.to_json(),
            'friday': self.friday.to_json(),
            'saturday': self.saturday.to_json(),
            'sunday': self.sunday.to_json()
        }


@dataclass
class AlarmDay:
    """
    Alarm config for a single day.

    :var active: If the alarm is active on this day.
    :var time: The time the alarm should go off.
    :var override_active: If the alarm should only once go off at a different time.
    :var override_time: The time the alarm should go off if override_active is true.
    """
    active: bool
    time: time
    override_active: bool
    override_time: time

    @classmethod
    def from_json(cls, config_json: dict) -> AlarmDay:
        """Create a AlarmDay object from a dict."""
        return cls(
            config_json['active'],
            time.fromisoformat(config_json['time']),
            config_json['overrideActive'],
            time.fromisoformat(config_json['overrideTime'])
        )

    def to_json(self) -> dict:
        """Dump this AlarmDay object into a dict."""
        return {
            'active': self.active,
            'time': self.time.isoformat(),
            'overrideActive': self.override_active,
            'overrideTime': self.override_time.isoformat()
        }

File: src/web/src/lib/app-state/parts/config/alarm-config.ts
import {debug} from "debug";
import {getSnapshot, type Instance, types} from "mobx-state-tree";
import {DateTime} from "luxon";

const log = debug("weckPiWeb:appState:config:alarm");

/** The ISO-8601 weekday number, where 1 is Monday and 7 is Sunday. */
type IsoWeekday = 1 | 2 | 3 | 4 | 5 | 6 | 7;

/** Custom MST type for luxon's DateTime object that serializes to the ISO-8601 time format. */
const Time = types.custom<string, DateTime>({
    name: "Time",
    fromSnapshot(snapshot: string): DateTime {
        return DateTime.fromISO(snapshot);
    },
    toSnapshot(value: DateTime): string {
        return value.toISOTime({includeOffset: false})!;
    },
    isTargetType(value: DateTime | string): boolean {
        return value instanceof DateTime;
    },
    getValidationMessage(snapshot: string): string {
        const dt = DateTime.fromISO(snapshot);
        return dt.isValid ? "" : dt.invalidExplanation!;
    }
});

const AlarmDay = types.model({
    active: types.boolean,
    time: Time,
    overrideActive: types.boolean,
    overrideTime: Time
});

export const AlarmConfig = types.model({
    monday: AlarmDay,
    tuesday: AlarmDay,
    wednesday: AlarmDay,
    thursday: AlarmDay,
    friday: AlarmDay,
    saturday: AlarmDay,
    sunday: AlarmDay
}).views(self => ({
    getAlarmConfig(weekday: IsoWeekday) {
        switch (weekday) {
            case 1:
                return self.monday;
            case 2:
                return self.tuesday;
            case 3:
                return self.wednesday;
            case 4:
                return self.thursday;
            case 5:
                return self.friday;
            case 6:
                return self.saturday;
            case 7:
                return self.sunday;
            default:
                throw new Error("Invalid weekday");
        }
    }
})).views(self => ({
    /** Get the alarm that would go off next. */
    getNextAlarm() {
        // Get a list of all weekdays beginning with the current weekday
        const currentTime = DateTime.local().set({
            day: self.getAlarmConfig(1).time.day,
            year: self.getAlarmConfig(1).time.year,
            month: self.getAlarmConfig(1).time.month
        });
        const currentWeekday = (DateTime.local().weekday as IsoWeekday);

        const weekdays = ([1, 2, 3, 4, 5, 6, 7]
            .slice(currentWeekday - 1)
            .concat([1, 2, 3, 4, 5, 6, 7]
                .slice(0, currentWeekday - 1)) as IsoWeekday[]);

        log("currentTime=%s, currentWeekday=%s, weekdays=%o", currentTime, currentWeekday, weekdays);

        // Find the first weekday with an active alarm
        for (const weekday of weekdays) {
            const alarmConfig: Instance<typeof AlarmDay> = self.getAlarmConfig(weekday);
            log("weekday=%s, alarmConfig=%o", weekday, getSnapshot(alarmConfig));

            // Check if the alarm is active and, if it's today's alarm, check if the alarm is still in the future
            if (alarmConfig.overrideActive && (weekday !== currentWeekday || alarmConfig.overrideTime > currentTime)) {
                return {time: alarmConfig.overrideTime, weekday: weekday};
            }

            // Same for the normal alarm time
            if (alarmConfig.active && (weekday !== currentWeekday || alarmConfig.time > currentTime)) {
                return {time: alarmConfig.time, weekday: weekday};
            }
        }

        return null;
    }
})).actions(self => ({
    setActive(weekday: IsoWeekday, value: boolean) {
        self.getAlarmConfig(weekday).active = value;
    },
    setTime(weekday: IsoWeekday, value: DateTime) {
        self.getAlarmConfig(weekday).time = value;
    },
    setOverrideActive(weekday: IsoWeekday, value: boolean) {
        self.getAlarmConfig(weekday).overrideActive = value;
    },
    setOverrideTime(weekday: IsoWeekday, value: DateTime) {
        self.getAlarmConfig(weekday).overrideTime = value;
    }
}));
